File: Semester9/AI/final/gca.py
from typing import List, Tuple
from functools import reduce
from itertools import product

def is_possible(graph: List[List[int]], colors: List[int], vertex: int, color: int) -> bool:
    """Check if it's valid to assign the given color to the vertex."""
    return all(graph[vertex][i] == 0 or colors[i] != color for i in range(len(graph)))

def solve(graph: List[List[int]], num_colors: int, colors: List[int], vertex: int) -> List[int]:
    """Recursively solve the graph coloring problem."""
    if vertex == len(graph):
        return colors
    
    for color in range(1, num_colors + 1):
        if is_possible(graph, colors, vertex, color):
            colors[vertex] = color
            result = solve(graph, num_colors, colors.copy(), vertex + 1)
            if result:
                return result
            colors[vertex] = 0
    
    return []

def graph_color(graph: List[List[int]], num_colors: int) -> Tuple[bool, List[int]]:
    """Main function to solve the graph coloring problem."""
    colors = [0] * len(graph)
    result = solve(graph, num_colors, colors, 0)
    return bool(result), result

def display_solution(colors: List[int]) -> None:
    """Display the coloring solution."""
    text_colors = ["", "RED", "GREEN", "BLUE", "YELLOW", "ORANGE", "PINK",
                  "BLACK", "BROWN", "WHITE", "PURPLE", "VIOLET"]
    print("\nColors:", " ".join(text_colors[color] for color in colors))

def main():
    print("Graph Coloring Algorithm Test\n")
    
    # Get number of vertices
    V = int(input("Enter number of vertices\n"))
    
    # Get graph matrix
    print("\nEnter matrix\n")
    graph = [[int(x) for x in input().split()] for _ in range(V)]
    
    # Get number of colors
    num_colors = int(input("\nEnter number of colors\n"))
    
    # Solve the graph coloring problem
    success, solution = graph_color(graph, num_colors)
    
    if success:
        print("\nSolution exists")
        display_solution(solution)
    else:
        print("No solution")

if __name__ == "__main__":
    main()

File: Semester9/AI/final/gca.js
// Graph Coloring Algorithm Implementation
const COLORS = {
    1: "RED",
    2: "GREEN",
    3: "BLUE",
    4: "YELLOW",
    5: "ORANGE",
    6: "PINK",
    7: "BLACK",
    8: "BROWN",
    9: "WHITE",
    10: "PURPLE",
    11: "VIOLET"
};

// Pure function to check if a color can be assigned to a vertex
const isColorPossible = (graph, colors, vertex, color) => {
    return !graph[vertex].some((edge, i) => edge === 1 && colors[i] === color);
};

// Pure function to display the solution
const displaySolution = (colors) => {
    console.log("\nColors:", colors.map(c => COLORS[c]).join(" "));
};

// Main graph coloring function using functional approach
const graphColor = (graph, numColors) => {
    const V = graph.length;
    const colors = new Array(V).fill(0);

    // Recursive function to solve the coloring problem
    const solve = (vertex) => {
        // Base case: all vertices are colored
        if (vertex === V) {
            return true;
        }

        // Try all colors for current vertex
        for (let color = 1; color <= numColors; color++) {
            if (isColorPossible(graph, colors, vertex, color)) {
                colors[vertex] = color;
                
                // Recursively color remaining vertices
                if (solve(vertex + 1)) {
                    return true;
                }
                
                // Backtrack
                colors[vertex] = 0;
            }
        }
        
        return false;
    };

    // Attempt to solve the coloring problem
    const hasSolution = solve(0);
    
    if (hasSolution) {
        console.log("\nSolution exists");
        displaySolution(colors);
    } else {
        console.log("No solution");
    }
};

// Example usage:
const exampleGraph = [
    [0, 1, 1, 1],
    [1, 0, 1, 0],
    [1, 1, 0, 1],
    [1, 0, 1, 0]
];

// Test the algorithm
console.log("Graph Coloring Algorithm Test\n");
console.log("Example graph:");
console.log(exampleGraph.map(row => row.join(" ")).join("\n"));
console.log("\nAttempting to color with 3 colors:");
graphColor(exampleGraph, 3);
